types/validator_set.go

```diff
@@ -68,25 +78,64 @@ func (vals *ValidatorSet) IncrementAccum(times int) {
 		panic("Cannot call IncrementAccum with non-positive times")
 	}
 
-	// Add VotingPower * times to each validator and order into heap.
+	const shiftEveryNthIter = 10
+	var proposer *Validator
+	// call IncrementAccum(1) times times:
+	for i := 0; i < times; i++ {
+		shiftByAvgAccum := i%shiftEveryNthIter == 0
+		proposer = vals.incrementAccum(shiftByAvgAccum)
+	}
+	isShiftedAvgOnLastIter := (times-1)%shiftEveryNthIter == 0
```

Comment:
Yes, especially if `times` is large this will save some time computing / finishing `IncrementAccum`. It's an optimization. If you think that it sin't justified, I can remove that logic and simply average on each iteration.

If `times == 1`, the shift will happen in the for-loop (once) and not at the end of `IncrementAccum`.

ref https://github.com/tendermint/tendermint/pull/2785#discussion_r235043750